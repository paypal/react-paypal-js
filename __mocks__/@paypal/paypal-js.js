import { loadScript as loadScriptOriginal } from "@paypal/paypal-js";

export function loadScript(options) {
    return new Promise((resolve) => {
        loadScriptOriginal(options);
        process.nextTick(() => resolve({}));
    });
}
