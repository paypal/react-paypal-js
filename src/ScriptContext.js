import React, {createContext, useContext, useEffect, useReducer} from 'react';
import PropTypes from 'prop-types';
import {loadScript} from '@paypal/paypal-js';

const ScriptContext = createContext();
const ScriptDispatchContext = createContext();

function scriptReducer(state, action) {
    switch (action.type) {
        case 'setLoadingState':
            return {
                options: {
                    ...state.options,
                },
                isLoaded: action.value,
            };
        case 'changeCurrency':
            return {
                options: {
                    ...state.options,
                    currency: action.value,
                },
                isLoaded: false,
            };

        default: {
            throw new Error(`Unhandled action type: ${action.type}`);
        }
    }
}

function useScriptReducer() {
    const scriptContext = useContext(ScriptContext);
    const dispatchContext = React.useContext(ScriptDispatchContext);
    if (scriptContext === undefined || dispatchContext === undefined) {
        throw new Error(
            'useScriptReducer must be used within a ScriptProvider'
        );
    }
    return [scriptContext, dispatchContext];
}

function ScriptProvider({options, children}) {
    const initialState = {
        options,
        isLoaded: false,
    };

    const [state, dispatch] = useReducer(scriptReducer, initialState);

    useEffect(() => {
        if (state.isLoaded) return;

        let isSubscribed = true;
        loadScript(state.options).then(() => {
            if (isSubscribed) {
                dispatch({type: 'setLoadingState', value: true});
            }
        });
        return () => {
            isSubscribed = false;
        };
    });

    return (
        <ScriptContext.Provider value={state}>
            <ScriptDispatchContext.Provider value={dispatch}>
                {children}
            </ScriptDispatchContext.Provider>
        </ScriptContext.Provider>
    );
}

ScriptProvider.propTypes = {
    children: PropTypes.any,
    options: PropTypes.shape({
        'client-id': PropTypes.string.isRequired,
    }),
};

export {ScriptProvider, useScriptReducer};
