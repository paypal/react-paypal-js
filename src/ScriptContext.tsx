import React, {
    createContext,
    useContext,
    useEffect,
    useReducer,
    FunctionComponent,
} from "react";
import { loadScript } from "@paypal/paypal-js";
import type { PayPalScriptOptions } from "@paypal/paypal-js/types/script-options";

export interface ReactPayPalScriptOptions extends PayPalScriptOptions {
    "data-react-paypal-script-id": string;
}

const SCRIPT_LOADING_STATE = {
    INITIAL: "initial",
    PENDING: "pending",
    REJECTED: "rejected",
    RESOLVED: "resolved",
} as const;

type ScriptLoadingState =
    typeof SCRIPT_LOADING_STATE[keyof typeof SCRIPT_LOADING_STATE];

interface ScriptContextState {
    options: ReactPayPalScriptOptions;
    loadingStatus: ScriptLoadingState;
}

interface ScriptContextDerivedState {
    options: ReactPayPalScriptOptions;
    isInitial: boolean;
    isPending: boolean;
    isRejected: boolean;
    isResolved: boolean;
}

type ScriptReducerAction =
    | {
          type: "setLoadingStatus";
          value: ScriptLoadingState;
      }
    | {
          type: "resetOptions";
          value: PayPalScriptOptions | ReactPayPalScriptOptions;
      };

type ScriptReducerDispatch = (action: ScriptReducerAction) => void;

const ScriptContext = createContext<ScriptContextState | null>(null);
const ScriptDispatchContext = createContext<ScriptReducerDispatch | null>(null);

function scriptReducer(state: ScriptContextState, action: ScriptReducerAction) {
    switch (action.type) {
        case "setLoadingStatus":
            return {
                options: {
                    ...state.options,
                },
                loadingStatus: action.value,
            };
        case "resetOptions":
            // destroy existing script to make sure only one script loads at a time
            destroySDKScript(state.options["data-react-paypal-script-id"]);
            return {
                loadingStatus: SCRIPT_LOADING_STATE.PENDING,
                options: {
                    ...action.value,
                    "data-react-paypal-script-id": `${getNewScriptID()}`,
                },
            };

        default: {
            return state;
        }
    }
}

function getNewScriptID() {
    return `react-paypal-js-${Math.random().toString(36).substring(7)}`;
}

function destroySDKScript(reactPayPalScriptID: string) {
    const scriptNode = document.querySelector(
        `script[data-react-paypal-script-id="${reactPayPalScriptID}"]`
    );
    if (scriptNode === null) return;

    if (scriptNode.parentNode) {
        scriptNode.parentNode.removeChild(scriptNode);
    }
}

function usePayPalScriptReducer(): [
    ScriptContextDerivedState,
    ScriptReducerDispatch
] {
    const scriptContext = useContext(ScriptContext);
    const dispatchContext = useContext(ScriptDispatchContext);

    if (scriptContext === null || dispatchContext === null) {
        throw new Error(
            "usePayPalScriptReducer must be used within a PayPalScriptProvider"
        );
    }

    const { loadingStatus, ...restScriptContext } = scriptContext;

    const derivedStatusContext = {
        ...restScriptContext,
        isInitial: loadingStatus === SCRIPT_LOADING_STATE.INITIAL,
        isPending: loadingStatus === SCRIPT_LOADING_STATE.PENDING,
        isResolved: loadingStatus === SCRIPT_LOADING_STATE.RESOLVED,
        isRejected: loadingStatus === SCRIPT_LOADING_STATE.REJECTED,
    };

    return [derivedStatusContext, dispatchContext];
}

interface ScriptProviderProps {
    options: PayPalScriptOptions;
    children?: React.ReactNode;
    deferLoading?: boolean;
}

const PayPalScriptProvider: FunctionComponent<ScriptProviderProps> = ({
    options,
    children,
    deferLoading = false,
}: ScriptProviderProps) => {
    const initialState = {
        options: {
            ...options,
            "data-react-paypal-script-id": `${getNewScriptID()}`,
        },
        loadingStatus: deferLoading
            ? SCRIPT_LOADING_STATE.INITIAL
            : SCRIPT_LOADING_STATE.PENDING,
    };

    const [state, dispatch] = useReducer(scriptReducer, initialState);

    useEffect(() => {
        if (
            deferLoading === false &&
            state.loadingStatus === SCRIPT_LOADING_STATE.INITIAL
        ) {
            return dispatch({
                type: "setLoadingStatus",
                value: SCRIPT_LOADING_STATE.PENDING,
            });
        }

        if (state.loadingStatus !== SCRIPT_LOADING_STATE.PENDING) return;

        let isSubscribed = true;
        loadScript(state.options)
            .then(() => {
                if (isSubscribed) {
                    dispatch({
                        type: "setLoadingStatus",
                        value: SCRIPT_LOADING_STATE.RESOLVED,
                    });
                }
            })
            .catch(() => {
                if (isSubscribed) {
                    dispatch({
                        type: "setLoadingStatus",
                        value: SCRIPT_LOADING_STATE.REJECTED,
                    });
                }
            });
        return () => {
            isSubscribed = false;
        };
    }, [options, deferLoading, state.loadingStatus]);

    return (
        <ScriptContext.Provider value={state}>
            <ScriptDispatchContext.Provider value={dispatch}>
                {children}
            </ScriptDispatchContext.Provider>
        </ScriptContext.Provider>
    );
};

export { PayPalScriptProvider, usePayPalScriptReducer };
