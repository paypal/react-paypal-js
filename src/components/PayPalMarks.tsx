import React, { useEffect, useRef, useState, FC, ReactNode } from "react";
import { usePayPalScriptReducer } from "../hooks/scriptProviderHooks";
import { getPayPalWindowNamespace, generateErrorMessage } from "../utils";
import { DATA_NAMESPACE } from "../constants";
import type {
    PayPalMarksComponentOptions,
    PayPalMarksComponent,
} from "@paypal/paypal-js/types/components/marks";

export interface PayPalMarksComponentProps extends PayPalMarksComponentOptions {
    /**
     * Pass a css class to the div container.
     */
    className?: string;
    children?: ReactNode;
}

/**
The `<PayPalMarks />` component is used for conditionally rendering different payment options using radio buttons.
The [Display PayPal Buttons with other Payment Methods guide](https://developer.paypal.com/docs/business/checkout/add-capabilities/buyer-experience/#display-paypal-buttons-with-other-payment-methods) describes this style of integration in detail.
It relies on the `<PayPalScriptProvider />` parent component for managing state related to loading the JS SDK script.
```jsx
    <PayPalMarks />
```
This component can also be configured to use a single funding source similar to the [standalone buttons](https://developer.paypal.com/docs/business/checkout/configure-payments/standalone-buttons/) approach.
A `FUNDING` object is exported by this library which has a key for every available funding source option.
```jsx
    import { PayPalScriptProvider, PayPalMarks, FUNDING } from "@paypal/react-paypal-js";
    <PayPalScriptProvider options={{ "client-id": "test", components: "buttons,marks" }}>
        <PayPalMarks fundingSource={FUNDING.PAYPAL}/>
    </PayPalScriptProvider>
```
*/
export const PayPalMarks: FC<PayPalMarksComponentProps> = ({
    className = "",
    children,
    ...markProps
}: PayPalMarksComponentProps) => {
    const [{ isResolved, options }] = usePayPalScriptReducer();
    const markContainerRef = useRef<HTMLDivElement>(null);
    const [isEligible, setIsEligible] = useState(true);
    const [, setErrorState] = useState(null);

    /**
     * Render PayPal Mark into the DOM
     */
    const renderPayPalMark = (mark: PayPalMarksComponent) => {
        const { current } = markContainerRef;

        // only render the mark when eligible
        if (!current || !mark.isEligible()) {
            return setIsEligible(false);
        }
        // Remove any children before render it again
        if (current.firstChild) {
            current.removeChild(current.firstChild);
        }

        mark.render(current).catch((err) => {
            // component failed to render, possibly because it was closed or destroyed.
            if (current === null || current.children.length === 0) {
                // paypal marks container is no longer in the DOM, we can safely ignore the error
                return;
            }
            // paypal marks container is still in the DOM
            setErrorState(() => {
                throw new Error(
                    `Failed to render <PayPalMarks /> component. ${err}`
                );
            });
        });
    };

    useEffect(() => {
        // verify the sdk script has successfully loaded
        if (isResolved === false) return;

        const paypalWindowNamespace = getPayPalWindowNamespace(
            options[DATA_NAMESPACE]
        );

        // verify dependency on window object
        if (
            paypalWindowNamespace === undefined ||
            paypalWindowNamespace.Marks === undefined
        ) {
            const { components, [DATA_NAMESPACE]: dataNamespace } = options;

            return setErrorState(() => {
                throw new Error(
                    generateErrorMessage({
                        componentName: PayPalMarks.displayName as string,
                        requiredOption: "marks",
                        components,
                        dataNamespace,
                    })
                );
            });
        }

        renderPayPalMark(paypalWindowNamespace.Marks({ ...markProps }));
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isResolved, markProps.fundingSource]);

    return (
        <>
            {isEligible ? (
                <div ref={markContainerRef} className={className} />
            ) : (
                children
            )}
        </>
    );
};

PayPalMarks.displayName = "PayPalMarks";
