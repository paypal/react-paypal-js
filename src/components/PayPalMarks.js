import React, { useEffect, useRef } from "react";
import PropTypes from "prop-types";
import { useScriptReducer } from "../ScriptContext";
/**
 * The `<PayPalMarks />` component is used for conditionally rendering different payment options using radio buttons. The [Using Radio Buttons](https://developer.paypal.com/docs/checkout/integration-features/mark-flow/) guide describes this style of integration in detail.
 * It relies on the `<PayPalScriptProvider />` parent component for managing state related to loading the JS SDK script.
 *
 * ```jsx
 *     <PayPalMarks />
 * ```
 *
 * This component can also be configured to use a single funding source similar to the [standalone buttons](https://developer.paypal.com/docs/checkout/integration-features/standalone-buttons/) approach.
 * A `FUNDING` object is exported by this library which has a key for every available funding source option.
 *
 * ```js
 *     import { FUNDING } from '@paypal/react-paypal-js'
 * ```
 *
 * Use this `FUNDING` constant to set the `fundingSource` prop.
 *
 * ```jsx
 *     <PayPalMarks fundingSource={FUNDING.PAYPAL}/>
 * ```
 */
export default function Marks(props) {
    const [{ isLoaded, options }] = useScriptReducer();
    const markContainerRef = useRef(null);
    const mark = useRef(null);

    useEffect(() => {
        if (isLoaded && !mark.current) {
            verifyGlobalStateForMarks(options);

            mark.current = window.paypal.Marks({ ...props });

            if (mark.current.isEligible()) {
                mark.current.render(markContainerRef.current);
            }
        }
    });

    return <div ref={markContainerRef} />;
}

function verifyGlobalStateForMarks({ components = "" }) {
    if (typeof window.paypal.Marks !== "undefined") {
        return;
    }

    let errorMessage =
        "Unable to render <PayPalMarks /> because window.paypal.Marks is undefined.";

    // the JS SDK does not load the Marks component by default. It must be passed into the "components" query parameter.
    if (!components.includes("marks")) {
        const expectedComponents = components ? `${components},marks` : "marks";

        errorMessage +=
            "\nTo fix the issue, add 'marks' to the list of components passed to the parent PayPalScriptProvider:" +
            `\n\`<PayPalScriptProvider options={{ components: '${expectedComponents}'}}>\`.`;
    }
    throw new Error(errorMessage);
}

Marks.propTypes = {
    /**
     * The individual mark to render. Use the `FUNDING` constant exported by this library to set this value. The full list can be found [here](https://developer.paypal.com/docs/checkout/integration-features/standalone-buttons/#complete-your-integration).
     */
    fundingSource: PropTypes.string,
};
