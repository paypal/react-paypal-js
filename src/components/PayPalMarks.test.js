import React from "react";
import { render, waitFor } from "@testing-library/react";

import { PayPalScriptProvider, usePayPalScriptReducer } from "../ScriptContext";
import PayPalMarks from "./PayPalMarks";
import { FUNDING } from "@paypal/sdk-constants";

describe("<PayPalMarks />", () => {
    beforeEach(() => {
        window.paypal = {};

        jest.spyOn(console, "error");
        console.error.mockImplementation(() => {});
    });
    afterEach(() => {
        console.error.mockRestore();
    });

    test("should pass props to window.paypal.Marks()", async () => {
        window.paypal.Marks = () => {
            return {
                isEligible: jest.fn(),
                render: jest.fn(),
            };
        };

        const spyOnMarks = jest.spyOn(window.paypal, "Marks");

        render(
            <PayPalScriptProvider
                options={{ "client-id": "sb", components: "marks" }}
            >
                <PayPalMarks fundingSource={FUNDING.CREDIT} />
            </PayPalScriptProvider>
        );

        await waitFor(() =>
            expect(spyOnMarks).toHaveBeenCalledWith({
                fundingSource: FUNDING.CREDIT,
            })
        );
    });

    test("should throw an error when no components are passed to the PayPalScriptProvider", () => {
        expect(() =>
            render(
                <PayPalScriptProvider options={{ "client-id": "sb" }}>
                    <SetIsLoadedToTrue />
                    <PayPalMarks />
                </PayPalScriptProvider>
            )
        ).toThrowErrorMatchingSnapshot();
    });

    test("should throw an error when the 'marks' component is missing from the components list passed to the PayPalScriptProvider", () => {
        expect(() =>
            render(
                <PayPalScriptProvider
                    options={{
                        "client-id": "sb",
                        components: "buttons,messages",
                    }}
                >
                    <SetIsLoadedToTrue />
                    <PayPalMarks />
                </PayPalScriptProvider>
            )
        ).toThrowErrorMatchingSnapshot();
    });
});

// immediately sets the PayPalScriptProvider `isLoaded` state to true
function SetIsLoadedToTrue() {
    const [, dispatch] = usePayPalScriptReducer();
    dispatch({ type: "setIsLoaded", value: true });
    return null;
}
