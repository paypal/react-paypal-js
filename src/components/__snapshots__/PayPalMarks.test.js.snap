// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PayPalMarks /> should throw an error when no components are passed to the PayPalScriptProvider 1`] = `
"Unable to render <PayPalMarks /> because window.paypal.Marks is undefined.
To fix the issue, add 'marks' to the list of components passed to the parent PayPalScriptProvider:
\`<PayPalScriptProvider options={{ components: 'marks'}}>\`."
`;

exports[`<PayPalMarks /> should throw an error when the 'marks' component is missing from the components list passed to the PayPalScriptProvider 1`] = `
"Unable to render <PayPalMarks /> because window.paypal.Marks is undefined.
To fix the issue, add 'marks' to the list of components passed to the parent PayPalScriptProvider:
\`<PayPalScriptProvider options={{ components: 'buttons,messages,marks'}}>\`."
`;
